llvm/include/llvm/CodeGen/DroppedVariableStats.h

```diff
@@ -0,0 +1,216 @@
+
+  /// A DenseSet tracking whether a scope was visited before.
+  DenseSet<const DIScope *> VisitedScope;
+  /// A stack of DenseMaps, which map the name of an llvm::Function to a
+  /// DenseMap of VarIDs and their inlinedAt locations before an optimization
+  /// pass has run.
+  SmallVector<DenseMap<StringRef, DenseMap<VarID, DILocation *>>> InlinedAts;
+  /// Remove a dropped #dbg_value VarID from all Sets in the
+  /// DroppedVariablesBefore stack.
+  void removeVarFromAllSets(VarID Var, const Function *F) {
+    // Do not remove Var from the last element, it will be popped from the
+    // stack.
+    for (auto &DebugVariablesMap : llvm::drop_end(DebugVariablesStack))
+      DebugVariablesMap[F].DebugVariablesBefore.erase(Var);
+  }
+  /// Return true if \p Scope is the same as \p DbgValScope or a child scope of
+  /// \p DbgValScope, return false otherwise.
+  bool isScopeChildOfOrEqualTo(const DIScope *Scope,
+                               const DIScope *DbgValScope);
+  /// Return true if \p InlinedAt is the same as \p DbgValInlinedAt or part of
+  /// the InlinedAt chain, return false otherwise.
+  bool isInlinedAtChildOfOrEqualTo(const DILocation *InlinedAt,
+                                   const DILocation *DbgValInlinedAt);
+
+  /// Calculate the number of dropped variables in an llvm::Function or
+  /// llvm::MachineFunction and print the relevant information to stdout.
+  void calculateDroppedStatsAndPrint(DebugVariables &DbgVariables,
+                                     StringRef FuncName, StringRef PassID,
+                                     StringRef FuncOrModName,
+                                     StringRef PassLevel, const Function *Func);
+
+  /// Check if a \p Var has been dropped or is a false positive.
+  bool wasDropped(DILocation *DbgLoc, const DIScope *Scope,
+                  const DIScope *DbgValScope,
+                  DenseMap<VarID, DILocation *> &InlinedAtsMap, VarID Var,
+                  unsigned &DroppedCount);
+  /// Run code to populate relevant data structures over an llvm::Function or
+  /// llvm::MachineFunction.
+  void run(DebugVariables &DbgVariables, StringRef FuncName, bool Before);
+  /// Populate the VarIDSet and InlinedAtMap with the relevant information
+  /// needed for before and after pass analysis to determine dropped variable
+  /// status.
+  void populateVarIDSetAndInlinedMap(
+      const DILocalVariable *DbgVar, DebugLoc DbgLoc, DenseSet<VarID> &VarIDSet,
+      DenseMap<StringRef, DenseMap<VarID, DILocation *>> &InlinedAtsMap,
+      StringRef FuncName, bool Before);
+  /// Visit every VarID in the \p DebugVariablesBeforeSet and check if it may
+  /// have been dropped by an optimization pass.
+  virtual void visitEveryDebugVariable(
+      unsigned &DroppedCount, DenseSet<VarID> &DebugVariablesBeforeSet,
+      DenseSet<VarID> &DebugVariablesAfterSet,
+      DenseMap<VarID, DILocation *> &InlinedAtsMap, VarID Var) = 0;
```

Comment:
I think this comment and function name could be improved for clarity.

The comment makes it sound like this function is visiting every VarID, but its that this function is called with every VarID. Maybe this is just a hangover from a previous iteration, since it also looks like `DebugVariablesBeforeSet` isn't used in the virtual override (neither IR nor MIR in the next patch)?

Reply:
Oh you're right, this is my bad, yes originally the function would visit every VarID, but it doesn't anymore because of some small changes I made. Let me update this